packages/twenty-server/src/workspace/workspace-query-builder/factories/find-duplicates-query.factory.ts

```diff
@@ -0,0 +1,140 @@
+
+import { FieldsStringFactory } from './fields-string.factory';
+
+@Injectable()
+export class FindDuplicatesQueryFactory {
+  private readonly logger = new Logger(FindDuplicatesQueryFactory.name);
+
+  constructor(
+    private readonly fieldsStringFactory: FieldsStringFactory,
+    private readonly argsAliasFactory: ArgsAliasFactory,
+  ) {}
+
+  async create<Filter extends RecordFilter = RecordFilter>(
+    args: FindDuplicatesResolverArgs<Filter>,
+    options: WorkspaceQueryBuilderOptions,
+    currentRecord?: Record<string, unknown>,
+  ) {
+    const fieldsString = await this.fieldsStringFactory.create(
+      options.info,
+      options.fieldMetadataCollection,
+      options.objectMetadataCollection,
+    );
+
+    const argsData = this.getFindDuplicateBy<Filter>(
+      args,
+      options,
+      currentRecord,
+    );
+
+    const duplicateCondition = this.buildDuplicateCondition(
+      options.objectMetadataItem,
+      argsData,
+      args.id,
+    );
+
+    const filters = stringifyWithoutKeyQuote(duplicateCondition);
+
+    return `
+      query {
+        ${computeObjectTargetTable(options.objectMetadataItem)}Collection${
+          filters ? `(filter: ${filters})` : ''
+        } {
+          ${fieldsString}
+        }
+      }
+    `;
+  }
+
+  getFindDuplicateBy<Filter extends RecordFilter = RecordFilter>(
+    args: FindDuplicatesResolverArgs<Filter>,
+    options: WorkspaceQueryBuilderOptions,
+    currentRecord?: Record<string, unknown>,
+  ) {
+    if (currentRecord) {
+      return currentRecord;
+    }
+
+    return this.argsAliasFactory.create(
+      args.data ?? {},
+      options.fieldMetadataCollection,
+    );
+  }
+
+  buildQueryForExistingRecord(
+    id: string,
+    options: WorkspaceQueryBuilderOptions,
+  ) {
+    return `
+      query {
+        ${computeObjectTargetTable(
+          options.objectMetadataItem,
+        )}Collection(filter: { id: { eq: "${id}" }}){
+          edges {
+            node {
+              ${this.getApplicableDuplicateCriteriaCollection(
+                options.objectMetadataItem,
+              )
+                .flatMap((dc) => dc.columnNames)
+                .join('\n')}
+            }
+          }
+        }
+      }
+    `;
+  }
+
+  private buildDuplicateCondition(
+    objectMetadataItem: ObjectMetadataInterface,
+    argsData?: Record<string, unknown>,
+    filteringByExistingRecordId?: string,
+  ) {
+    if (!argsData) {
+      return;
+    }
+
+    const criteriaCollection =
+      this.getApplicableDuplicateCriteriaCollection(objectMetadataItem);
+
+    const criteriaWithMatchingArgs = criteriaCollection.filter((criteria) =>
+      criteria.columnNames.every((columnName) => !!argsData[columnName]),
+    );
+
+    const filterCriteria = criteriaWithMatchingArgs.map((criteria) =>
+      Object.fromEntries(
+        criteria.columnNames.map((col) => [
```

Comment:
```suggestion
        criteria.columnNames.map((columnName) => [
```

Better not to use abbreviation :)